crates/solver/src/interactions/allowances.rs

```diff
@@ -135,45 +156,66 @@ impl AllowanceManager {
 #[async_trait::async_trait]
 impl AllowanceManaging for AllowanceManager {
     async fn get_allowances(&self, tokens: HashSet<H160>, spender: H160) -> Result<Allowances> {
-        Ok(Allowances::new(
-            spender,
-            fetch_allowances(self.web3.clone(), tokens, self.owner, spender).await?,
-        ))
+        Ok(fetch_allowances(
+            self.web3.clone(),
+            self.owner,
+            hashmap! { spender => tokens },
+        )
+        .await?
+        .remove(&spender)
+        .unwrap_or_else(|| Allowances::empty(spender)))
     }
 
-    async fn get_approval(&self, token: H160, spender: H160, amount: U256) -> Result<Approval> {
-        self.get_allowances(hashset![token], spender)
-            .await?
-            .approve_token(token, amount)
+    async fn get_approvals(&self, requests: &[ApprovalRequest]) -> Result<Vec<Approval>> {
+        let mut spender_tokens = HashMap::<_, HashSet<_>>::new();
+        for request in requests {
+            spender_tokens
+                .entry(request.spender)
+                .or_default()
+                .insert(request.token);
+        }
+
+        let allowances = fetch_allowances(self.web3.clone(), self.owner, spender_tokens).await?;
+        requests
+            .iter()
+            .map(|request| {
+                allowances
+                    .get(&request.spender)
+                    .with_context(|| {
+                        format!("no allowances found for spender {}", request.spender)
+                    })?
+                    .approve_token(request.token, request.amount)
+            })
+            .collect()
```

Comment:
I was very confused here. I didn't know [this](https://doc.rust-lang.org/std/result/enum.Result.html#impl-FromIterator%3CResult%3CA%2C%20E%3E%3E) was a thing.